plugin/execute_command.py

```diff
@@ -12,13 +13,27 @@ def __init__(self, view: sublime.View) -> None:
     def run(self,
             edit: sublime.Edit,
             command_name: Optional[str] = None,
-            command_args: Optional[Any] = None) -> None:
+            command_args: Optional[List[Any]] = None) -> None:
         client = self.client_with_capability('executeCommandProvider')
         if client and command_name:
             window = self.view.window()
             if window:
                 window.status_message("Running command {}".format(command_name))
-            self._send_command(client, command_name, command_args)
+            self._send_command(client, command_name, self._expand_variables(command_args))
+
+    def _expand_variables(self, command_args: List[Any]) -> List[Any]:
+        for i, arg in enumerate(command_args):
+            if arg == "${file}":
```

Comment:
It would be much nicer/easier to use `sublime.expand_variables` here. Something like this?

```python
variables = self.view.window().extract_variables()
region = self.view.sel()[-1]
variables['uri'] = uri_from_view(self.view)
variables['offset'] = region.b
variables['selection_begin'] = region.begin()
variables['selection_end'] = region.end()
variables['selection'] = self.view.substr(region)
return sublime.expand_variables(command_args, variables)
```

There are also already some concepts like Position and Range that should be included. `$position` and `$range`?

The `offset`, `selection_begin` and `selection_end` are UTF-32 offset-based, I don't think this will mesh well with the protocol, which generally uses a (row, col_utf16) pair to denote positions in a document. What server are you using?

Reply:
Oh, I thought the mapping for `expand_variables()` must contain predefined strings. I will try your suggestion, which seems more elegant.

> The offset, selection_begin and selection_end are UTF-32 offset-based

I indeed saw that for the example `a𐐀b` in the specs, this implementation gives the wrong offset value. However, I guess it is impossible to solve with the current ST API. I quickly experimented with something like
```python
from .core.protocol import Point
from .core.views import point_to_offset

point_to_offset(Point(*self.view.rowcol(self.view.sel()[-1].b)), self.view)
```
but this seems useless since it only converts back and forth. Do you know a solution?

I use the Julia server, which expects a single int offset: https://github.com/julia-vscode/LanguageServer.jl/blob/cdb566d369389a43c1cdf88f3d7d1e69127dd73e/src/requests/actions.jl#L38

Reply:
The question becomes whether Julia server then expects UTF-32 or UTF-16 offset. But communicating with UTF-16 offsets is madness, I suppose the only sane choice is UTF-32.

On the other hand, it *should* ask for a JSON dict `{"line": x, "character": y}` which is standard LSP.

> However, I guess it is impossible to solve with the current ST API. 

It is solved in the st4000-exploration branch with the new API.

Reply:
I could not get `sublime.expand_variables()` to work like that, as it throws an error if the values for `variables` are not strings. Seems like this function is only for simple string replacement, and doesn't work with numbers, objects or evaluated functions.

I added `"${position}"` and `"${range}"` from the LSP specs.

> It is solved in the st4000-exploration branch with the new API.

I think it should work for most cases now, and ST3 users could at least profit from this implementation. Maybe it could be adjusted for the ST4 branch later then.

Reply:
Also I noticed that we should not send `"arguments": null` if no arguments are provided.

Reply:
Hmm, it's working here for me. Most likely SublimeHQ added this functionality for ST4.

```
>>> sublime.expand_variables(['$folder'], window.extract_variables())
['/home/raoul/.config/sublime-text-3/Packages/LSP']
>>> sublime.expand_variables({'foo': '$platform'}, window.extract_variables())
{'foo': 'Linux'}
```

Reply:
Yes, but these two examples are both strings. You cannot add you own objects to `variables` and cannot evaluate functions for it.

Reply:
```
>>> sublime.expand_variables('$test', {'test': 42})
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Program Files\Sublime Text 3\sublime.py", line 221, in expand_variables
    return sublime_api.expand_variables(val, variables)
TypeError: String required
```
And at least `view.substr()` didn't work for me when used as a value in `variables`.

Reply:
Thanks for explaining the problem. You're correct in that we can't use sublime.expand_variables for this.

But this makes me question whether using this string-template notation is a good idea. The current impl does not recognize e.g. that `$offset` and `${offset}` are the same thing. Can we instead use something like a dict to denote the type/kind of argument? For instance:

```
[{"kind": "position"}, {"kind": "selection"}, {"kind": "constant", "value": 5}]
```
would define a command signature with three arguments, the first arg is the position, the second arg is the selection, and the third arg is the constant integer 5.

Reply:
If you want to go far with this, you could also have a `{"kind": "input-text"}` that would show a TextInputHandler for the user to type in some text, and a `{"kind": "input-list", "value": ["option1, "option2"]}` that would show a ListInputHandler with the choices option1 and option2.

Reply:
> The current impl does not recognize e.g. that $offset and ${offset} are the same thing.

No big problem, I could use `if arg in ['$offset', '${offset}']` then. The only thing that wouldn't work with this implementation is an argument like `'some text ${file_uri} more text'`, but this doesn't make much sense anyway, as most variables are replaced by an int or dict and I don't think there are many use cases for something like this example.

The third argument in
```
[{"kind": "position"}, {"kind": "selection"}, {"kind": "constant", "value": 5}]
```
doesn't make much sense, as it's the same as
```
[{"kind": "position"}, {"kind": "selection"}, 5]
```
but actually in my opinion this notation would only make it more complicated for users, and the `$offset` or `${offset}` notation is already a standard in many Sublime config files. If we want, we could additionally use `sublime.expand_variables()` for the other standard window variables like `$folder`, but I'm not sure if there are any use cases for these as well. If I should change the notation to `{"kind": "...."}`, I at least would like to see a third opinion who prefers that too.

Reply:
I like the `input-text` and `-list` idea (preferably `"$input_text"`/`"${input_text}"` 😉 ) but there might be some uncertainties with how it should work:
* for multiple arguments like these, just show multiple quick panels?
* should the input be treated as a string - that would make it impossible to input numbers

Maybe we should better leave this for another pull request.